File: app/src/main/java/com/example/armonitor/MainActivity.java
package com.example.armonitor;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.Gravity;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;


import org.eclipse.paho.android.service.MqttAndroidClient;
import org.eclipse.paho.client.mqttv3.IMqttActionListener;
import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
import org.eclipse.paho.client.mqttv3.IMqttToken;
import org.eclipse.paho.client.mqttv3.MqttCallback;
import org.eclipse.paho.client.mqttv3.MqttClient;
import org.eclipse.paho.client.mqttv3.MqttException;
import org.eclipse.paho.client.mqttv3.MqttMessage;


public class MainActivity extends AppCompatActivity {

    private Button toAr;
    private Button mqttConnect;
    private static final String TAG = MainActivity.class.getSimpleName();



    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        toAr = findViewById(R.id.toAr);
        toAr.setOnClickListener(new View.OnClickListener() { //BUTTON FOR OPENING AR ACTIVITY
            @Override
            public void onClick(View v) {
                openArActivity();
            }
        });

        final String topic = "topic/testing";
        final int qos = 1;
        String clientId = MqttClient.generateClientId();
        final MqttAndroidClient client =
                new MqttAndroidClient(this.getApplicationContext(), "tcp://10.202.50.252:1883",
                        clientId);






        //CONNECT////////////////////////////////////////////////////////////////////////////////////////////
        try {
            IMqttToken token = client.connect();
            token.setActionCallback(new IMqttActionListener() {
                @Override
                public void onSuccess(IMqttToken asyncActionToken) {
                    // We are connected
                    Log.d(TAG, "onSuccess");

                    //SUBSCRIBE: MUST BE INSIDE THE onSuccess() method after client.connect()//////////////////
                    try {
                        final IMqttToken subToken = client.subscribe(topic, qos);
                        subToken.setActionCallback(new IMqttActionListener() {
                            @Override
                            public void onSuccess(IMqttToken asyncActionToken) {
                                // The message was published
                                Log.d(TAG, "Subscribed");

                            }

                            @Override
                            public void onFailure(IMqttToken asyncActionToken,
                                                  Throwable exception) {
                                // The subscription could not be performed, maybe the user was not
                                // authorized to subscribe on the specified topic e.g. using wildcards
                                Log.d(TAG, "Subscribe failed");

                            }
                        });
                    } catch (MqttException e) {
                        e.printStackTrace();
                    }
                    //SUBSCRIBE*/////////////////////////////////////////////////////////////////
                }

                @Override
                public void onFailure(IMqttToken asyncActionToken, Throwable exception) {
                    // Something went wrong e.g. connection timeout or firewall problems
                    Log.d(TAG, "onFailure");

                }
            });
        } catch (MqttException e) {
            e.printStackTrace();
        }
        //CONNECT*////////////////////////////////////////////////////////////////////////////////


        client.setCallback(new MqttCallback() {
            @Override
            public void connectionLost(Throwable cause) {
                Log.d(TAG, "Connection lost");
            }

            @Override
            public void messageArrived(String topic, MqttMessage message) throws Exception {
                String payload = message.toString();
                Log.d(TAG, payload);
                Toast.makeText(getApplicationContext(), payload, payload.length()).show();
                DisplayMessage(payload);

            }

            @Override
            public void deliveryComplete(IMqttDeliveryToken token) {
                Log.d(TAG, "Delivery complete");
            }
        });



    }

    //METHOD FOR OPENING AR ACTIVITY
    private void openArActivity() {
        Intent openArView = new Intent(this, ArActivity.class);
        startActivity(openArView);
    }

    private void DisplayMessage(String msg)
    {
        LinearLayout ll = (LinearLayout) findViewById(R.id.receivedLayout);
        TextView mqttMsg = new TextView( getApplicationContext());
        mqttMsg.setBackgroundColor(getResources().getColor(R.color.lime));
        mqttMsg.setGravity(Gravity.CENTER | Gravity.BOTTOM);
        mqttMsg.setText(msg);
        ll.addView(mqttMsg);
    }
}

File: app/src/main/java/com/example/armonitor/MainActivity.kt
package com.example.armonitor


import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import org.eclipse.paho.android.service.MqttAndroidClient

class MainActivity : AppCompatActivity() {

    private var toAr: Button? = null
    private lateinit var mqttAndroidClient: MqttAndroidClient
    private val mqtt: MqttHandler = MqttHandler(this)
    val topics: Array<String> = arrayOf("temp", "paper")
    val qos = 1
    val red = Color.parseColor("#F7370D")
    val green = Color.parseColor("#9EDC0B")

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val toAr = findViewById<Button>(R.id.toAr)

        toAr.setOnClickListener(View.OnClickListener
        {
            openArActivity()
        })


        mqtt.connect(topics)
        mqtt.receiveMessages(::DisplayMessage)
    }

    ///METHOD FOR OPENING AR ACTIVITY//////////////////////////////////////////
    private fun openArActivity() {
        val openArView = Intent(this, ArActivity::class.java)
        startActivity(openArView)
    }

    fun DisplayMessage(msg: String, topic: String) {


        var ll = findViewById<View>(R.id.receivedLayout) as LinearLayout
        val params = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)
        val currentView = ll.findViewWithTag<TextView>(topic)
        params.setMargins(ViewGroup.LayoutParams.MATCH_PARENT, 10, ViewGroup.LayoutParams.MATCH_PARENT, 10)

        if (currentView == null) {
            val mqttMsg = TextView(applicationContext)
            mqttMsg.layoutParams = params
            mqttMsg.setBackgroundColor(setColor(topic, msg.toInt()))
            mqttMsg.gravity = Gravity.CENTER or Gravity.BOTTOM
            mqttMsg.setTextSize(30.0F)
            mqttMsg.setTag(topic)
            mqttMsg.text = topic + "       " + msg

            ll.addView(mqttMsg)
        } else {
            currentView.setBackgroundColor(setColor(topic, msg.toInt()))
            currentView.text = topic + "      " + msg
        }
    }

    fun setColor(topic: String, value: Int): Int {

        val maxTemp = 50
        val minPaper = 100
        if (topic >= "temp") {
            if (value > maxTemp)
            {
                return red
            }
            else
            {
                return green
            }
        }

        else if (topic == "paper") {
            if (value < minPaper)
            {
                return red
            }

            else
            {
                return green
            }
        }
        return 0
    }

    companion object {
        private val TAG = MainActivity::class.java.simpleName
    }
}


